### PR TITLE
WIP SR-1715: require superclass to have a designated initializer

### PR DESCRIPTION
**Pull request by randomstep**
# Description 
After checking the superclass in other ways, this PR verifies that the superclass has at least one designated initializer. Classes like NSProxy do not have a designated init, and thus are not sub-classable / instantiatable from Swift. The current error is lacking, and this PR provides a more targeted error.

# Bug Link
Resolves [SR-1715](https://bugs.swift.org/browse/SR-1715).

# Questions before running CI
Before CI is run, I'd like any PR feedback. Specifically:
- [ ] The tests seem clearly in the wrong place to me. Something like "super_constructor.swift" seems like a better place, but is not importing Foundation. I could add it, or I could add a new test file. Not sure I should. Will look for more init tests that my tests might better fit.
- [ ] while fixing tests, an apparently unrelated test started failing. It has to do with checking sil instructions. I may have changed the `A` class in the test version of Foundation (by adding an init), and that may have changed some assumptions. I am not sure where to even start figuring this out. It does appear to be Swift 2 syntax still in the test, so I could at least clean that up (duplicate labels and similar).
- [ ] A number of tests needed updates as they failed this new check. I added inits to some non-Swift objects, and expected errors in some other places. Currently not sure how to handle the circular dependency test failure though. They clearly should show the user the circular errors, not the less-specific (and un-useful) designated-init error. I was unable to fix one of the circularity errors. I'd love to know that a type was part of a cycle, in which case this new check could be skipped entirely. Maybe more of the super checks should be skipped in that case? As a worst-case, might it be acceptable to document the failing test to also expect this new error? My take: we should avoid that if at all feasible.
- Style: the bools in this file are mixed on whether they start upper or lower case. I tried to match the style my code was closest to, so my bools are camelCase. Should they be TitleCase instead?
- I also looked at adding a related test in "accessibility.swift" (code below). But after some thought, I think that is more in the court of [SR-2484](https://bugs.swift.org/browse/SR-2484) and should probably be a separate PR at a minimum. Let me know if you think this should be covered under this PR instead. Here is that test code:
```swift
// SR-1715: no available designated initializers
class Base {
	private init() {}
}

class SubClass: Base { // expected-error {{'cannot inherit from class 'Base' because it has no designated initializers'}}
}
```
- as a separate commit, I'm also considering refactoring the entire visit function this lives inside. Probably not necessary or desired, but if I'm wrong, let me know. (Things like de-nesting from the ifs with inverse statements, extracting some of the large work chunks to separate functions.)

### CODE REVIEW

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -4787,6 +4787,34 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
           }
         }
 
+        // Require the superclass to have an available designated initializer,
+        // unless we have already determined the class or superclass is invalid.
+        if (!isInvalidSuperclass && !Super->isInvalid()) {
+          bool superIsConstructible = false; // can be satisfied by convenience initializers
+          bool superHasDesignatedInitializer = false; // is subclassable
+          for (auto member : Super->getMembers()) {
```
Comment (jeremy-w) [inline lib/Sema/TypeCheckDecl.cpp]: [Turn predicate loops into predicate functions?](https://llvm.org/docs/CodingStandards.html#turn-predicate-loops-into-predicate-functions)

Comment (randomstep) [reply lib/Sema/TypeCheckDecl.cpp]: Might work. I'd like to find a refactor that cleans this up for sure.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2002,6 +2002,9 @@ ERROR(inheritance_from_cf_class,none,
 ERROR(inheritance_from_objc_runtime_visible_class,none,
       "cannot inherit from class %0 because it is only visible via the "
       "Objective-C runtime", (Identifier))
+ERROR(inheritance_from_class_with_no_designated_initializers,none,
+	  "cannot inherit from class %0 because it has no designated initializers",
```
Comment (xwu) [inline include/swift/AST/DiagnosticsSema.def]: Drive-by nit: spaces, not tabs, are used in this file; please align this line as shown above.

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -4787,6 +4787,34 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
           }
         }
 
+        // Require the superclass to have an available designated initializer,
+        // unless we have already determined the class or superclass is invalid.
+        if (!isInvalidSuperclass && !Super->isInvalid()) {
+          bool superIsConstructible = false; // can be satisfied by convenience initializers
```
Comment (xwu) [inline lib/Sema/TypeCheckDecl.cpp]: Another nit: all comments are capitalized and punctuated, and this file respects the 80-character limit for each line.

**Review by xwu — COMMENTED** [review]
Hurray! Just some formatting comments.

**Comment by CodaFi** [comment]
@randomstep Are you still interested in this patch?  If so, please rebase it and address the review comments.  Otherwise, we can close it out and I can update the SR to point another contributor at this work to see if they can take it over the goal line.

Thanks!

**Comment by CodaFi** [comment]
Alright, there hasn't been any movement or response here in a while. I'm going to close this out and discharge the SR assignment. If you wish to pursue this change further, please reopen this pull request, rebase it, and address the review feedback.

Thanks!
